linked_list/linked_list.py

```diff
@@ -1,79 +1,134 @@
 
 # Defines a node in the singly linked list
 class Node:
-
     def __init__(self, value, next_node = None):
         self.value = value
         self.next = next_node
 
 # Defines the singly linked list
 class LinkedList:
     def __init__(self):
-      self.head = None # keep the head private. Not accessible outside this class
+        self.head = None # keep the head private. Not accessible outside this class
 
     # returns the value in the first node
     # returns None if the list is empty
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def get_first(self):
-        pass
+        if self.head == None:
+            return None
+        else:
+            return self.head.value
 
 
     # method to add a new node with the specific data value in the linked list
     # insert the new node at the beginning of the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def add_first(self, value):
-        pass
+        new_node = Node(value, next_node=self.head)
+        self.head = new_node
 
     # method to find if the linked list contains a node with specified value
     # returns true if found, false otherwise
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def search(self, value):
-        pass
+        current = self.head
+        while current:
+            if current.value == value:
+                return True 
+            current = current.next
+        return False 
 
     # method that returns the length of the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def length(self):
-        pass
+        current_node = self.head
+        count = 0
+        while current_node != None:
+            count +=1
+            current_node = current_node.next
+        return count
 
     # method that returns the value at a given index in the linked list
     # index count starts at 0
     # returns None if there are fewer nodes in the linked list than the index value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def get_at_index(self, index):
-        pass
+        current_in = 0
+        current_node = self.head
+        if self.length() < index:

```

Comment:
Because the last index is length -1
```suggestion
        if self.length() <= index:

```